Filtering RU alleles

rnmitchell opened the pull request:
This PR incorporates filtering using detection thresholds, analytical thresholds and stutter thresholds. This PR will only address the RU allele; LUSPlus allele filtering will be included in a later PR. 

This PR also produces output files for use in either EuroForMix or STRmix.

rnmitchell commented on .github/workflows/cibuild.yml:
```diff
@@ -5,7 +5,7 @@ on: [push]
   build:
 
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-20.04
```
I ran into an issue with the CI failing all of the sudden (worked just fine the previous day). After googling, it looks like pinning a version of ubuntu solved the issue with python 3.6 [(link here)](https://github.com/slackapi/bolt-python/pull/776). If this is an issue, lmk and we can look for another work around.

rnmitchell replied:
Ok, I thought it may be something in GitHub. I clicked around to see if I could find it but couldn't. Thanks!

standage replied:
In Settings, Branch protection rules: I changed required from [3.6, 3.7] to [3.7, 3.8, 3.9]

rnmitchell commented:
@standage this is now ready for review. It's a lot, so if you want to go over anything in person, just let me know!

standage commented on lusSTR/cli.py:
```diff
@@ -105,16 +105,55 @@ def snps_subparser(subparsers):
+        'STRmix. If separate files are specified for EFM, the sample ID will be used as the '
+        'filename. Output files are in CSV format.'
+    )
+    cli.add_argument(
+        '--profile-type', dest='profile', choices=['evidence', 'reference'], default='evidence',
+        help='Choose the type of profile, either evidence or reference. Default is evidence.'
+    )
```
Leaving this question here now, in case I don't find the answer while looking at the rest of the code: how are evidence and reference profiles treated differently? Anything in addition to assuming a single diploid contributor for references?

rnmitchell replied:
The output format (for both STRmix and EFM) is the only thing that differs (references only report two alleles and do not report the # of reads).

standage commented on lusSTR/filter.py:
```diff
@@ -0,0 +1,191 @@
+    return final_df, flags_df
+
+
+def EFM_output(df, outfile, profile, separate=False):
+    if outfile is None:
+        outfile = sys.stdout
+    if profile == 'reference':
+        infile = df[df.allele_type == 'real_allele']
+    else:
+        infile = df[df.allele_type != 'noise']
+    infile_sort = infile.sort_values(by=['SampleID', 'Locus', 'RU_Allele'], ascending=True)
```
I might want to have you walk me through this function.

standage commented on lusSTR/filter.py:
```diff
@@ -0,0 +1,191 @@
+                df_complete.iloc[:, :4].to_csv(f'{name}_reference.csv', index=False)
+
+
+def STRmix_output(df, outdir, profile):
+    if profile == 'reference':
+        infile = df[df.allele_type == 'real_allele']
+    else:
+        infile = df
+    data_combine = infile.groupby(
```
And this one as well.

standage commented on lusSTR/filters.json:
```diff
@@ -0,0 +1,488 @@
+{
+    "CSF1PO": {
+        "MinimumNumberReadsForDynamicThresholds": 650,
+        "DetectionThresholdStaticCount": 10,
+        "DetectionThresholdDynamicPercent": 0,
+        "DetectionThresholdUse": "Static",
+        "AnalyticalThresholdStaticCount": 20,
```
Did you intend to include two copies of this file?

rnmitchell replied:
lol nope. I don't even know how I did that.

standage commented on lusSTR/filter_settings.py:
```diff
@@ -0,0 +1,297 @@
+    filter_marker_data = json.load(fh)
+
+
+def filters(data_order, locus, total_reads):
+    metadata = filter_marker_data[locus]
+    if len(data_order) == 1:
+        if thresholds('Detection', metadata, total_reads, data_order['Reads'][0])[1] is False:
+            data_order = pd.DataFrame()
+        elif thresholds('Analytical', metadata, total_reads, data_order['Reads'][0])[1] is False:
+            data_order[['allele_type', 'perc_noise']] = ['noise', 1.0]
+        elif thresholds('Analytical', metadata, total_reads, data_order['Reads'][0])[1] is True:
+            data_order['allele_type'] = 'real_allele'
```
This function is...high maintenance. :smiley:

On one hand, the tests are passing, so I don't want to hold up progress on this PR too much. On the other hand, I worry about adding support for filtering LUS/LUS+ and SNP alleles with this as the foundation.

I've often been ok with not being able to grasp all the details when doing code reviews for lusSTR. But if you're agreeable, perhaps we can schedule some time to walk through this code in person (or in real time remotely) next week.

standage reviewed (commented):
Good job on getting this code working @rnmitchell! I don't have any glaring concerns, but I think I need a bit of help parsing some of the code. When you have some availability, let's discuss the code and see if we can identify a few low-hanging-fruit targets for refactoring, either in this PR or a dedicated subsequent PR.

rnmitchell commented on filters.json:
```diff
@@ -0,0 +1,407 @@
+{
+    "CSF1PO": {
+        "MinimumNumberReadsForDynamicThresholds": 650,
+        "DetectionThresholdStaticCount": 10,
+        "DetectionThresholdDynamicPercent": 0,
+        "DetectionThresholdUse": "Static",
```
Yes, these are our thresholds. I should make it more clear that this file should be modified for other labs and their established thresholds.

rnmitchell commented:
@standage I've added additional functionalities since you've looked at this PR. I've added in the ability to create STRmix output files for NGS data (i.e. using the sequence strings from lusSTR) and perform filtering using detection thresholds, analytical thresholds and same size thresholds. Same size thresholds are when two sequences differ but are the same CE allele (sequence with highest number of reads are compared to other sequences, those with reads < threshold are dropped). Stutter filters are not yet implemented for sequence strings (much more complicated).

standage commented on lusSTR/filter_settings.py:
```diff
@@ -0,0 +1,343 @@
+    filter_marker_data = json.load(fh)
+
+
+def filters(data_order, locus, total_reads, datatype):
+    metadata = filter_marker_data[locus]
+    if len(data_order) == 1:
+        if thresholds('Detection', metadata, total_reads, data_order['Reads'][0])[1] is False:
+            data_order = pd.DataFrame()
+        elif thresholds('Analytical', metadata, total_reads, data_order['Reads'][0])[1] is False:
+            data_order[['allele_type', 'perc_noise']] = ['noise', 1.0]
+        elif thresholds('Analytical', metadata, total_reads, data_order['Reads'][0])[1] is True:
+            data_order['allele_type'] = 'real_allele'
```
I think this function is my biggest remaining source of concern. It's doing a lot of things, and I'm wondering if it can be decomposed into 2-4 smaller functions: one that handles `len(data_order) == 1`, one that handles `datatype == "ce"`, and so on.

Also, `data_order` is the allele read count information, right? Another pass over this code to revisit variable names would probably improve legibility quite a bit.

rnmitchell replied:
ok I'll take a look

rnmitchell replied:
I did some refactoring and renaming variables to hopefully make things a little more clear. Let me know when you get a chance to look @standage and if it's up to your standard 😆

standage replied:
Lol, just so we're clear, I don't think *I* code up to "my standard" very frequently. I hope this review has been helpful, or at least that it increases the likelihood that someone (possibly one of us) looking at this code 3 years will make sense of it!

rnmitchell replied:
I'm kidding! Hence the laughing face. The Standage standard is great!

rnmitchell replied:
And much needed for my code 😄

standage replied:
The code is much clearer now. Thanks for humoring me!

rnmitchell replied:
Thank you for your help!!
